src/NuGet.Core/NuGet.Configuration/Settings/SettingsFile.cs

```diff
@@ -0,0 +1,240 @@
+            {
+                throw new ArgumentNullException(nameof(fileName), Resources.Argument_Cannot_Be_Null_Or_Empty);
+            }
+
+            if (!FileSystemUtility.IsPathAFile(fileName))
+            {
+                throw new ArgumentException(Resources.Settings_FileName_Cannot_Be_A_Path, nameof(fileName));
+            }
+
+            DirectoryPath = directoryPath;
+            FileName = fileName;
+            IsMachineWide = isMachineWide;
+
+            XDocument config = null;
+            ExecuteSynchronized(() =>
+            {
+                config = XmlUtility.GetOrCreateDocument(CreateDefaultConfig(), ConfigFilePath);
+            });
+
+            _xDocument = config;
+
+            _rootElement = new NuGetConfiguration(_xDocument.Root, origin: this);
+        }
+
+        /// <summary>
+        /// Gets the section with a given name.
+        /// </summary>
+        /// <param name="sectionName">name to match sections</param>
+        /// <returns>null if no section with the given name was found</returns>
+        public SettingSection GetSection(string sectionName)
+        {
+            return _rootElement.GetSection(sectionName);
+        }
+
+        /// <summary>
+        /// Adds or updates the given <paramref name="item"/> to the settings.
+        /// </summary>
+        /// <param name="sectionName">section where the <paramref name="item"/> has to be added. If this section does not exist, one will be created.</param>
+        /// <param name="item">item to be added to the settings.</param>
+        /// <returns>true if the item was successfully updated or added in the settings</returns>
+        public void AddOrUpdate(string sectionName, SettingItem item)
+        {
+            _rootElement.AddOrUpdate(sectionName, item);
+        }
+
+        /// <summary>
+        /// Removes the given <paramref name="item"/> from the settings.
+        /// If the <paramref name="item"/> is the last item in the section, the section will also be removed.
+        /// </summary>
+        /// <param name="sectionName">Section where the <paramref name="item"/> is stored. If this section does not exist, the method will throw</param>
+        /// <param name="item">item to be removed from the settings</param>
+        /// <remarks> If the SettingsFile is a machine wide config this method will throw</remarks>
+        public void Remove(string sectionName, SettingItem item)
+        {
+            _rootElement.Remove(sectionName, item);
+        }
+
+        /// <summary>
+        /// Flushes any in-memory updates in the SettingsFile to disk.
+        /// </summary>
+        public void SaveToDisk()
+        {
+            if (IsDirty)
+            {
+                ExecuteSynchronized(() =>
+                {
+                    FileSystemUtility.AddFile(ConfigFilePath, _xDocument.Save);
+                });
+
+                IsDirty = false;
+                SettingsChanged?.Invoke(this, EventArgs.Empty);
```

Comment:
shouldn't this SettingChanged event be raised only once per `SaveToDisk` in `Settings.cs`?

Reply:
You could subscribe to this `SettingsChange` if you want to know if a specific settingsFile changed. Settings subscribe to this event on each settings file. You are right that on settings I need to make sure this event shouldn't trigger the Settings event more than once. Good catch!